File: routes/contacts.js
const express = require('express');
const { body, validationResult, query } = require('express-validator');
const Contact = require('../models/Contact');
const { auth, adminAuth } = require('../middleware/auth');

const router = express.Router();

// @route   GET /api/contacts
// @desc    Get all contacts with filtering and pagination
// @access  Private
router.get('/', auth, [
  query('page').optional().isInt({ min: 1 }).withMessage('Page must be a positive integer'),
  query('limit').optional().isInt({ min: 1, max: 100 }).withMessage('Limit must be between 1 and 100'),
  query('type').optional().isIn(['supplier', 'customer', 'vendor', 'distributor']),
  query('search').optional().isString()
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const page = parseInt(req.query.page) || 1;
    const limit = parseInt(req.query.limit) || 20;
    const skip = (page - 1) * limit;

    // Build filter object
    const filter = { isActive: true };
    
    if (req.query.type) {
      filter.type = req.query.type;
    }
    
    if (req.query.search) {
      filter.$or = [
        { name: new RegExp(req.query.search, 'i') },
        { company: new RegExp(req.query.search, 'i') },
        { 'contactInfo.email': new RegExp(req.query.search, 'i') }
      ];
    }

    const contacts = await Contact.find(filter)
      .sort({ name: 1 })
      .skip(skip)
      .limit(limit);

    const total = await Contact.countDocuments(filter);

    res.json({
      contacts,
      pagination: {
        current: page,
        pages: Math.ceil(total / limit),
        total,
        hasNext: page < Math.ceil(total / limit),
        hasPrev: page > 1
      }
    });
  } catch (error) {
    console.error('Get contacts error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

// @route   GET /api/contacts/:id
// @desc    Get single contact
// @access  Private
router.get('/:id', auth, async (req, res) => {
  try {
    const contact = await Contact.findById(req.params.id);

    if (!contact) {
      return res.status(404).json({ message: 'Contact not found' });
    }

    res.json(contact);
  } catch (error) {
    console.error('Get contact error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

// @route   POST /api/contacts
// @desc    Create new contact
// @access  Private (Admin/Manager)
router.post('/', adminAuth, [
  body('name').notEmpty().withMessage('Name is required'),
  body('type').isIn(['supplier', 'customer', 'vendor', 'distributor']).withMessage('Invalid contact type'),
  body('contactInfo.email').optional().isEmail().withMessage('Please include a valid email'),
  body('contactInfo.phone').optional().isString(),
  body('businessInfo.creditLimit').optional().isFloat({ min: 0 }).withMessage('Credit limit must be a positive number'),
  body('businessInfo.paymentTerms').optional().isIn(['net_15', 'net_30', 'net_45', 'net_60', 'cash_on_delivery', 'prepaid']).withMessage('Invalid payment terms')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const contact = new Contact(req.body);
    await contact.save();

    res.status(201).json(contact);
  } catch (error) {
    console.error('Create contact error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

// @route   PUT /api/contacts/:id
// @desc    Update contact
// @access  Private (Admin/Manager)
router.put('/:id', adminAuth, [
  body('name').optional().notEmpty().withMessage('Name cannot be empty'),
  body('type').optional().isIn(['supplier', 'customer', 'vendor', 'distributor']).withMessage('Invalid contact type'),
  body('contactInfo.email').optional().isEmail().withMessage('Please include a valid email'),
  body('businessInfo.creditLimit').optional().isFloat({ min: 0 }).withMessage('Credit limit must be a positive number'),
  body('businessInfo.paymentTerms').optional().isIn(['net_15', 'net_30', 'net_45', 'net_60', 'cash_on_delivery', 'prepaid']).withMessage('Invalid payment terms')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const contact = await Contact.findById(req.params.id);
    if (!contact) {
      return res.status(404).json({ message: 'Contact not found' });
    }

    Object.assign(contact, req.body);
    await contact.save();

    res.json(contact);
  } catch (error) {
    console.error('Update contact error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

// @route   DELETE /api/contacts/:id
// @desc    Delete contact (soft delete)
// @access  Private (Admin/Manager)
router.delete('/:id', adminAuth, async (req, res) => {
  try {
    const contact = await Contact.findById(req.params.id);
    if (!contact) {
      return res.status(404).json({ message: 'Contact not found' });
    }

    contact.isActive = false;
    await contact.save();

    res.json({ message: 'Contact deleted successfully' });
  } catch (error) {
    console.error('Delete contact error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

// @route   GET /api/contacts/suppliers
// @desc    Get all suppliers
// @access  Private
router.get('/type/suppliers', auth, async (req, res) => {
  try {
    const suppliers = await Contact.find({
      type: 'supplier',
      isActive: true
    })
    .select('name company contactInfo specialties rating')
    .sort({ name: 1 });

    res.json(suppliers);
  } catch (error) {
    console.error('Get suppliers error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

// @route   GET /api/contacts/customers
// @desc    Get all customers
// @access  Private
router.get('/type/customers', auth, async (req, res) => {
  try {
    const customers = await Contact.find({
      type: 'customer',
      isActive: true
    })
    .select('name company contactInfo address')
    .sort({ name: 1 });

    res.json(customers);
  } catch (error) {
    console.error('Get customers error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

// @route   POST /api/contacts/:id/rate
// @desc    Rate a contact
// @access  Private
router.post('/:id/rate', auth, [
  body('rating').isInt({ min: 1, max: 5 }).withMessage('Rating must be between 1 and 5')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const contact = await Contact.findById(req.params.id);
    if (!contact) {
      return res.status(404).json({ message: 'Contact not found' });
    }

    contact.rating = req.body.rating;
    contact.lastContactDate = new Date();
    await contact.save();

    res.json(contact);
  } catch (error) {
    console.error('Rate contact error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

module.exports = router;

File: flask-app/routes/contacts.py
from flask import Blueprint, request, jsonify
from marshmallow import Schema, fields, ValidationError
from sqlalchemy import or_
from models.contact import Contact
from middleware.auth import auth_required, admin_required
from database import db

contacts_bp = Blueprint('contacts', __name__)

# Validation schemas
class ContactSchema(Schema):
    name = fields.Str(required=True)
    company = fields.Str()
    type = fields.Str(required=True, validate=lambda x: x in ['supplier', 'customer', 'vendor', 'distributor'])
    email = fields.Email()
    phone = fields.Str()
    mobile = fields.Str()
    fax = fields.Str()
    street = fields.Str()
    city = fields.Str()
    state = fields.Str()
    zip_code = fields.Str()
    country = fields.Str()
    tax_id = fields.Str()
    license_number = fields.Str()
    website = fields.Str()
    credit_limit = fields.Float()
    payment_terms = fields.Str(validate=lambda x: x in ['net_15', 'net_30', 'net_45', 'net_60', 'cash_on_delivery', 'prepaid'])
    specialties = fields.List(fields.Str())
    rating = fields.Int(validate=lambda x: 1 <= x <= 5)
    notes = fields.Str()

@contacts_bp.route('/', methods=['GET'])
@auth_required
def get_contacts():
    """Get all contacts with filtering and pagination"""
    try:
        # Get query parameters
        page = request.args.get('page', 1, type=int)
        limit = min(request.args.get('limit', 20, type=int), 100)
        contact_type = request.args.get('type')
        search = request.args.get('search')
        
        # Build query
        query = Contact.query.filter_by(is_active=True)
        
        if contact_type:
            query = query.filter_by(type=contact_type)
        
        if search:
            search_term = f"%{search}%"
            query = query.filter(
                or_(
                    Contact.name.ilike(search_term),
                    Contact.company.ilike(search_term),
                    Contact.email.ilike(search_term)
                )
            )
        
        # Apply pagination
        contacts = query.order_by(Contact.name).paginate(
            page=page, per_page=limit, error_out=False
        )
        
        return jsonify({
            'contacts': [contact.to_dict() for contact in contacts.items],
            'pagination': {
                'current_page': page,
                'per_page': limit,
                'total': contacts.total,
                'pages': contacts.pages
            }
        })
        
    except Exception as e:
        return jsonify({'message': 'Server error'}), 500

@contacts_bp.route('/', methods=['POST'])
@auth_required
def create_contact():
    """Create a new contact"""
    schema = ContactSchema()
    
    try:
        data = schema.load(request.get_json())
    except ValidationError as err:
        return jsonify({'errors': err.messages}), 400
    
    try:
        contact = Contact(**data)
        db.session.add(contact)
        db.session.commit()
        
        return jsonify({
            'message': 'Contact created successfully',
            'contact': contact.to_dict()
        }), 201
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'message': 'Server error during contact creation'}), 500

@contacts_bp.route('/<int:contact_id>', methods=['GET'])
@auth_required
def get_contact(contact_id):
    """Get a specific contact by ID"""
    try:
        contact = Contact.query.get_or_404(contact_id)
        return jsonify({'contact': contact.to_dict()})
    except Exception as e:
        return jsonify({'message': 'Contact not found'}), 404

@contacts_bp.route('/<int:contact_id>', methods=['PUT'])
@auth_required
def update_contact(contact_id):
    """Update a contact"""
    schema = ContactSchema()
    
    try:
        data = schema.load(request.get_json(), partial=True)
    except ValidationError as err:
        return jsonify({'errors': err.messages}), 400
    
    try:
        contact = Contact.query.get_or_404(contact_id)
        
        # Update contact fields
        for key, value in data.items():
            setattr(contact, key, value)
        
        db.session.commit()
        
        return jsonify({
            'message': 'Contact updated successfully',
            'contact': contact.to_dict()
        })
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'message': 'Server error during contact update'}), 500

@contacts_bp.route('/<int:contact_id>', methods=['DELETE'])
@admin_required
def delete_contact(contact_id):
    """Delete a contact (soft delete)"""
    try:
        contact = Contact.query.get_or_404(contact_id)
        contact.is_active = False
        db.session.commit()
        
        return jsonify({'message': 'Contact deleted successfully'})
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'message': 'Server error during contact deletion'}), 500

@contacts_bp.route('/suppliers', methods=['GET'])
@auth_required
def get_suppliers():
    """Get all active suppliers"""
    try:
        suppliers = Contact.query.filter_by(type='supplier', is_active=True).order_by(Contact.name).all()
        return jsonify({
            'suppliers': [supplier.to_dict() for supplier in suppliers]
        })
    except Exception as e:
        return jsonify({'message': 'Server error'}), 500 